src/ngraph/runtime/executable.hpp

```diff
@@ -0,0 +1,101 @@
+public:
+    Executable(runtime::Backend* backend);
+    virtual ~Executable();
+
+    /// \deprecated use execute method
+    /// \param outputs vector of runtime::Tensor used as outputs
+    /// \param inputs vector of runtime::Tensor used as inputs
+    /// \returns true if iteration is successful, false otherwise
+    DEPRECATED virtual bool call(const std::vector<std::shared_ptr<runtime::Tensor>>& outputs,
+                                 const std::vector<std::shared_ptr<runtime::Tensor>>& inputs);
+
+    /// \brief Executes a single iteration of a Function.
+    /// \param outputs vector of runtime::Tensor used as outputs
+    /// \param inputs vector of runtime::Tensor used as inputs
+    /// \returns true if iteration is successful, false otherwise
+    virtual bool execute(const std::vector<runtime::Tensor*>& outputs,
+                         const std::vector<runtime::Tensor*>& inputs) = 0;
+
+    /// \brief Executes a single iteration of a Function.
+    /// \param outputs vector of runtime::Tensor used as outputs
+    /// \param inputs vector of runtime::Tensor used as inputs
+    /// \returns true if iteration is successful, false otherwise
+    DEPRECATED bool call_with_validate(const std::vector<std::shared_ptr<runtime::Tensor>>& outputs,
+                                       const std::vector<std::shared_ptr<runtime::Tensor>>& inputs);
+
+    /// \brief Validates and then Executes a single iteration of a Function.
+    /// \param outputs vector of runtime::Tensor used as outputs
+    /// \param inputs vector of runtime::Tensor used as inputs
+    /// \returns true if iteration is successful, false otherwise
+    bool validate_and_execute(const std::vector<runtime::Tensor*>& outputs,
+                              const std::vector<runtime::Tensor*>& inputs);
+
+    /// \brief Collect performance information gathered on a Function.
+    /// \returns Vector of PerformanceCounter information.
+    virtual std::vector<PerformanceCounter> get_performance_data() const;
+
+    /// \brief Validates a Function.
+    /// \param outputs vector of runtime::Tensor used as outputs
+    /// \param inputs vector of runtime::Tensor used as inputs
+    void validate(const std::vector<runtime::Tensor*>& outputs,
+                  const std::vector<runtime::Tensor*>& inputs);
+
+    /// \brief Query the input Parameters for a given Handle
+    /// \returns an ngraph::op::ParameterVector of all input parameters
+    const ngraph::ParameterVector& get_parameters() const;
+
+    /// \brief Query the output Results for a given Handle
+    /// \returns an ngraph::ResultVector of all input parameters
+    const ngraph::ResultVector& get_results() const;
+
+protected:
+    /// \brief Called at the end of compile to the the values to be returned by get_parameters
+    ///     and get_results
+    /// \param func The function with Results fully resolved.
+    void set_parameters_and_results(const Function& func);
+
+    runtime::Backend* get_backend();
```

Comment:
Just curious -- why is this here?  (It doesn't seem to be used by the base class; it seems reasonable to let the subclass track it iff the subclass needs it -- and then the subclass can get the type right.)

Reply:
There are some ops like FunctionCall that need to compile graphs. Since these ops are handled in the Executable I think I can remove this since the Executable is the object that does the compile.
It is currently only used in the INTERPRETER backend.

Reply:
<nods> In the PlaidML backend, there's a Build object that contains a pointer to a Compiler object, for exactly this sort of thing.

Reply:
I removed the need for get_backend() since the interpreter can just make a new executable internally when it is needed for ops like FunctionCall.